compat/win32/wsl.c

```diff
@@ -0,0 +1,109 @@
+#include "../../git-compat-util.h"
+#include "../win32.h"
+#include "config.h"
+#include "ntifs.h"
+#include "wsl.h"
+
+#define LX_FILE_METADATA_HAS_UID 0x1
+#define LX_FILE_METADATA_HAS_GID 0x2
+#define LX_FILE_METADATA_HAS_MODE 0x4
+#define LX_FILE_METADATA_HAS_DEVICE_ID 0x8
+#define LX_FILE_CASE_SENSITIVE_DIR 0x10
+typedef struct _FILE_STAT_LX_INFORMATION {
+	LARGE_INTEGER FileId;
+	LARGE_INTEGER CreationTime;
+	LARGE_INTEGER LastAccessTime;
+	LARGE_INTEGER LastWriteTime;
+	LARGE_INTEGER ChangeTime;
+	LARGE_INTEGER AllocationSize;
+	LARGE_INTEGER EndOfFile;
+	uint32_t FileAttributes;
+	uint32_t ReparseTag;
+	uint32_t NumberOfLinks;
+	ACCESS_MASK EffectiveAccess;
+	uint32_t LxFlags;
+	uint32_t LxUid;
+	uint32_t LxGid;
+	uint32_t LxMode;
+	uint32_t LxDeviceIdMajor;
+	uint32_t LxDeviceIdMinor;
+} FILE_STAT_LX_INFORMATION, *PFILE_STAT_LX_INFORMATION;
+typedef struct _FILE_FULL_EA_INFORMATION {
+	ULONG NextEntryOffset;
+	UCHAR Flags;
+	UCHAR EaNameLength;
+	USHORT EaValueLength;
+	CHAR EaName[1];
+} FILE_FULL_EA_INFORMATION, *PFILE_FULL_EA_INFORMATION;
+
+enum {
+	FileStatLxInformation = 70,
+};
+__declspec(dllimport) NTSTATUS WINAPI
+	NtQueryInformationFile(HANDLE FileHandle,
+			       PIO_STATUS_BLOCK IoStatusBlock,
+			       PVOID FileInformation, ULONG Length,
+			       uint32_t FileInformationClass);
+__declspec(dllimport) NTSTATUS WINAPI
+	NtSetInformationFile(HANDLE FileHandle, PIO_STATUS_BLOCK IoStatusBlock,
+			     PVOID FileInformation, ULONG Length,
+			     uint32_t FileInformationClass);
+__declspec(dllimport) NTSTATUS WINAPI
+	NtSetEaFile(HANDLE FileHandle, PIO_STATUS_BLOCK IoStatusBlock,
+		    PVOID EaBuffer, ULONG EaBufferSize);
+
+int write_wsl_stat(HANDLE hd, struct stat *st)
+{
+	char buf[0x40] = { 0 };
+	const char *names[] = { "$LXUID", "$LXGID", "$LXMOD" };
+	uint32_t vals[] = { (uint32_t)st->st_uid, (uint32_t)st->st_gid,
+			    st->st_mode };
+	char *endp = buf;
+	FILE_FULL_EA_INFORMATION *prev = 0;
+	IO_STATUS_BLOCK iob = { 0 };
+
+	assert(S_ISREG(st->st_mode) || S_ISDIR(st->st_mode));
+
+	for (int i = 0; i < sizeof(names)/sizeof(names[0]); ++i) {
+		FILE_FULL_EA_INFORMATION *p =
+			(FILE_FULL_EA_INFORMATION *)(((uintptr_t)endp + 3) &
+						     ~3llu);
+		char *vp;
+		if (prev)
+			prev->NextEntryOffset =
+				(ULONG)((char *)p - (char *)prev);
+		p->NextEntryOffset = 0;
+		p->Flags = 0;
+		p->EaNameLength = (unsigned char)strlen(names[i]);
+		p->EaValueLength = sizeof(vals[i]);
+		memcpy(p->EaName, names[i], (size_t)p->EaNameLength + 1);
+		vp = p->EaName + p->EaNameLength + 1;
+		memcpy(vp, &vals[i], sizeof(vals[i]));
+		endp = vp + sizeof(vals[i]);
+		prev = p;
+	}
```

Comment:
This loop looks quite complicated, _especially_ given that the keys all have the same length. It would probably be a lot simpler to extend the declaration of `struct _FILE_FULL_EA_INFORMATION` to hard-code the `EaName`s to a size of 7, and to also include the `uint32_t` value, and then have a simple array.

This _might_ need some `pragma pack` and/or `__attribute__(aligned(...))` games (see e.g. [here](https://stackoverflow.com/questions/14179748/what-is-the-difference-between-pragma-pack-and-attribute-aligned), and would make me more confident that the alignment is correct (rather than playing those `+ 3 & ~3llu` games). After that, it should be as easy as:

```c
FILE_FULL_EA_INFORMATION info[3] = { { 0 } };
strcpy(info[0].EaName, "$LXUID");
info[0].EaNameLength = 6;
info[0].value = uid;
strcpy(info[1].EaName, "$LXGID");
info[1].EaNameLength = 6;
info[1].value = gid;
strcpy(info[2].EaName, "$LXMOD");
info[2].EaNameLength = 6;
info[2].value = mode;
```

which I find a lot more readable/debuggable.